src/pkg/podman/error_test.go

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright © 2019 – 2021 Red Hat Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package podman
+
+import (
+	"bytes"
+	"errors"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestInternalError(t *testing.T) {
+	type input struct {
+		Msg string
+	}
+
+	type expect struct {
+		IsNil  bool
+		Error  string
+		Search string
+		Wrap   []string
+	}
+
+	testCases := []struct {
+		name   string
+		input  input
+		expect expect
+	}{
+		{
+			name: "Empty error message",
+			input: input{
+				Msg: "",
+			},
+			expect: expect{
+				IsNil: true,
+				Error: "",
+			},
+		},
+		{
+			name: "Text with no prolog before an error message",
+			input: input{
+				Msg: "There is no prolog before the error message.",
+			},
+			expect: expect{
+				Error: "There is no prolog before the error message.",
+			},
+		},
+		{
+			name: "Text with prolog before an error message",
+			input: input{
+				Msg: "There is the prolog.Error: an error message",
+			},
+			expect: expect{
+				Error:  "an error message",
+				Search: "an error message",
+			},
+		},
+		{
+			name: "Text with prolog before an error message (separated by newline)",
+			input: input{
+				Msg: "There is the prolog.\nError: an error message",
+			},
+			expect: expect{
+				Error:  "an error message",
+				Search: "an error message",
+			},
+		},
+		{
+			name: "Error message with several wrapped errors (prepended with \"Error:\")",
+			input: input{
+				Msg: "Error: level 1: level 2: level 3: level 4",
+			},
+			expect: expect{
+				Error:  "level 1: level 2: level 3: level 4",
+				Search: "level 4",
+				Wrap:   []string{"level 1", "level 2", "level 3", "level 4"},
+			},
+		},
+		{
+			name: "Error message with newline and with errors in \"bullet\" list",
+			input: input{
+				Msg: "Error: foobar:\n  err1\n  err2\n  err3",
+			},
+			expect: expect{
+				Error:  "foobar: err1: err2: err3",
+				Search: "err2",
+				Wrap:   []string{"foobar", "err1", "err2", "err3"},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			err := parseErrorMsg(bytes.NewBufferString(tc.input.Msg))
+
+			if tc.expect.IsNil {
+				assert.Nil(t, err)
+				return
+			} else {
+				assert.NotNil(t, err)
+			}
+
+			errInternal := err.(*internalError)
```

Comment:
I understand that the cast to an `internalError` is necessary to look at the slice of error strings and such, but shouldn't the public interface (ie., `Is`, `Unwrap`, etc.) be tested without the cast?

Reply:
The `internalError` type is not public and thus will never be used outside of the `podman` package. Casting to `error` to `*internalError` will be a common practice in the package.

Reply:
Also, `Is` and `Unwrap` are not a public interface in the classic `error` type, thus the cast is needed to invoke them.

Reply:
`Is` and `Unwrap` are not part of the `error` interface, but they are very much [public](https://blog.golang.org/go1.13-errors). That's how `errors.Unwrap` and `errors.Is` work. Errors created using the `%w` verb in `fmt.Errorf` have an `Unwrap` method. Etc..

The public interface is how almost every (all?) user of `internalError` will use it. So, we should use those in our test suite as much as possible. A cast to `internalError` should only be used if we are specifically looking inside the guts of the `internalError` implementation.